pkg/docker/docker.go

```diff
@@ -584,92 +605,97 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) (err error) {
 		return err
 	}
 
-	config := docker.Config{
-		Image: image,
-		User:  opts.User,
-	}
-
-	config, tarDestination := runContainerTar(opts, config, imageMetadata)
-
-	if opts.Env != nil {
-		config.Env = opts.Env
-	}
-	if opts.Stdin != nil {
-		config.OpenStdin = true
-		config.StdinOnce = true
-	}
-	if opts.Stdout != nil {
-		config.AttachStdout = true
-	}
-
-	ccopts := docker.CreateContainerOptions{Name: "", Config: &config}
-	ccopts.HostConfig = &docker.HostConfig{}
-	if opts.TargetImage {
-		ccopts.HostConfig.PublishAllPorts = true
-		ccopts.HostConfig.NetworkMode = opts.NetworkMode
-	} else if opts.NetworkMode != "" {
-		ccopts.HostConfig.NetworkMode = opts.NetworkMode
-	}
-
-	if opts.CGroupLimits != nil {
-		ccopts.HostConfig.Memory = opts.CGroupLimits.MemoryLimitBytes
-		ccopts.HostConfig.MemorySwap = opts.CGroupLimits.MemorySwap
-		ccopts.HostConfig.CPUShares = opts.CGroupLimits.CPUShares
-		ccopts.HostConfig.CPUQuota = opts.CGroupLimits.CPUQuota
-		ccopts.HostConfig.CPUPeriod = opts.CGroupLimits.CPUPeriod
-	}
-	ccopts.HostConfig.CapDrop = opts.CapDrop
-	glog.V(2).Infof("Creating container %s using config: %+v, hostconfig: %+v", ccopts.Name, ccopts.Config, ccopts.HostConfig)
+	cmd, tarDestination := runContainerTar(opts, imageMetadata)
+	createOpts.Config.Cmd = cmd
 
+	// Create a new container.
+	glog.V(2).Infof("Creating container with options {Name:%q Config:%+v HostConfig:%+v} ...", createOpts.Name, createOpts.Config, createOpts.HostConfig)
 	var container *docker.Container
 	if err := util.TimeoutAfter(DefaultDockerTimeout, func() error {
 		var createErr error
-		container, createErr = d.client.CreateContainer(ccopts)
+		container, createErr = d.client.CreateContainer(createOpts)
 		return createErr
 	}); err != nil {
 		return err
 	}
 
-	defer d.RemoveContainer(container.ID)
+	containerName := containerNameOrID(container)
+
+	// Container was created, so we defer its removal.
+	defer func() {
+		glog.V(4).Infof("Removing container %q ...", containerName)
+		if err := d.RemoveContainer(container.ID); err != nil {
+			glog.Warningf("Failed to remove container %q: %v", containerName, err)
+		} else {
+			glog.V(4).Infof("Removed container %q", containerName)
+		}
+	}()
 
-	glog.V(2).Infof("Attaching to container")
-	// creating / piping the channels in runContainerAttach lead to unintended hangs
-	attached := make(chan struct{})
-	wg := runContainerAttach(attached, container, opts, d)
-	attached <- <-attached
+	// Attach to the container.
+	glog.V(2).Infof("Attaching to container %q ...", containerName)
+	attachOpts := opts.asDockerAttachToContainerOptions()
+	attachOpts.Container = container.ID
+	if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
```

Comment:
> There might be goroutine leaking from go-dockerclient.

Or not. Since we always force remove the container, all the resources related to the attach call should be released.
I haven't checked the implementation of `client.hijack` enough to tell by heart if they have leaks.